File: Services/YouTubeAPIClient.swift
import Foundation

/// 動画一覧の1ページ分。
struct VideoPage {
    let items: [VideoItem]
    let nextPageToken: String?
}

/// YouTube Data API v3 クライアント。
/// スクレイピングは行わず、公式の Data API のみを使用する。
final class YouTubeAPIClient {

    private let session: URLSession
    private let baseURL = "https://www.googleapis.com/youtube/v3"
    /// 暴走防止のための最大ページ数（50件/ページ × 100 = 5000本）。
    private let maxPages = 100

    init(session: URLSession = .shared) {
        self.session = session
    }

    private func apiKey() throws -> String {
        guard let key = ConfigLoader.youtubeAPIKey() else {
            throw YouTubeAPIError.apiKeyMissing
        }
        return key
    }

    // MARK: - Public API

    /// 入力URL（または handle / channelId）からチャンネルを解決する。
    func resolveChannel(from inputURL: String) async throws -> Channel {
        let identifier = try ChannelResolver.parse(inputURL)
        switch identifier {
        case .channelId(let id):
            return try await fetchChannel(query: [("id", id)])
        case .handle(let handle):
            return try await fetchChannel(query: [("forHandle", "@\(handle)")])
        case .username(let name):
            return try await fetchChannel(query: [("forUsername", name)])
        case .customName(let name):
            let channelId = try await searchChannelId(byName: name)
            return try await fetchChannel(query: [("id", channelId)])
        case .video(let videoId):
            // 共有された動画URL → videos.list で投稿チャンネルを特定してから解決する（quota 1）。
            let channelId = try await fetchChannelId(forVideoId: videoId)
            return try await fetchChannel(query: [("id", channelId)])
        }
    }

    /// videoId からその動画を投稿したチャンネルの channelId を取得する（videos.list / quota 1）。
    func fetchChannelId(forVideoId videoId: String) async throws -> String {
        guard ChannelResolver.isVideoId(videoId) else {
            throw YouTubeAPIError.invalidVideoURL
        }
        let data = try await getData("videos", query: [("part", "snippet"), ("id", videoId)])
        return try Self.channelId(fromVideosListJSON: data)
    }

    /// videos.list のレスポンス JSON から channelId を取り出す（ネットワーク非依存＝テスト可能）。
    static func channelId(fromVideosListJSON data: Data) throws -> String {
        let response: VideoListResponse
        do {
            response = try JSONDecoder().decode(VideoListResponse.self, from: data)
        } catch {
            throw YouTubeAPIError.decodingError
        }
        guard let channelId = response.items.first?.snippet?.channelId,
              !channelId.isEmpty else {
            throw YouTubeAPIError.videoNotFound
        }
        return channelId
    }

    /// channelId から uploads プレイリストIDを取得する。
    func fetchUploadsPlaylistId(channelId: String) async throws -> String {
        let channel = try await fetchChannel(query: [("id", channelId)])
        guard let uploads = channel.uploadsPlaylistId else {
            throw YouTubeAPIError.uploadsPlaylistNotFound
        }
        return uploads
    }

    /// uploads プレイリストから全動画を取得し、古い順（publishedAt 昇順）で返す。
    /// uploads プレイリストは新しい順で返るため、古い順表示には全ページの取得が必要。
    func fetchVideos(playlistId: String) async throws -> [VideoItem] {
        var all: [VideoItem] = []
        var token: String? = nil
        var page = 0
        repeat {
            let result = try await fetchVideosPage(playlistId: playlistId, pageToken: token)
            all.append(contentsOf: result.items)
            token = result.nextPageToken
            page += 1
        } while token != nil && page < maxPages

        return all.sortedByPublishedDate(ascending: true)
    }

    /// 既に持っている動画に当たるまで、新しい順にページを取得して**新着だけ**返す。
    ///
    /// uploads プレイリストは新しい順に返るため、既知の動画に当たった時点で
    /// それ以降はすべて既知とみなせる。本数の多いチャンネルを毎回全件取り直さずに済む。
    ///
    /// - Returns: 新着（新しい順）と、既知の動画に到達したかどうか。
    ///   到達しなかった場合は差分が大きい（久しぶりに開いた等）ので、呼び出し側で全件取得に切り替える。
    func fetchNewVideos(playlistId: String,
                        knownVideoIds: Set<String>,
                        maxPages: Int = 5) async throws -> (videos: [VideoItem], reachedKnown: Bool) {
        var newItems: [VideoItem] = []
        var token: String?
        var page = 0
        var reachedKnown = false

        repeat {
            let result = try await fetchVideosPage(playlistId: playlistId, pageToken: token)
            for item in result.items {
                if knownVideoIds.contains(item.id) {
                    reachedKnown = true
                    break
                }
                newItems.append(item)
            }
            if reachedKnown { break }
            token = result.nextPageToken
            page += 1
        } while token != nil && page < maxPages

        // 最後まで見ても既知に当たらなかった＝そもそも全部が新しい（＝全件取得すべき）。
        return (newItems, reachedKnown || token == nil)
    }

    /// uploads プレイリストの1ページ分を取得する。
    func fetchVideosPage(playlistId: String, pageToken: String?) async throws -> VideoPage {
        var query: [(String, String)] = [
            ("part", "snippet,contentDetails"),
            ("playlistId", playlistId),
            ("maxResults", "50"),
        ]
        if let pageToken { query.append(("pageToken", pageToken)) }

        let response: PlaylistItemListResponse = try await get("playlistItems", query: query)
        let items: [VideoItem] = response.items.compactMap { item in
            guard let videoId = item.contentDetails?.videoId ?? item.snippet?.resourceId?.videoId else {
                return nil
            }
            let publishedString = item.contentDetails?.videoPublishedAt ?? item.snippet?.publishedAt
            let published = publishedString.flatMap(ISO8601.date(from:)) ?? Date.distantPast
            return VideoItem(
                id: videoId,
                title: item.snippet?.title ?? "(タイトルなし)",
                description: item.snippet?.description ?? "",
                publishedAt: published,
                thumbnailURL: item.snippet?.thumbnails?.bestURL,
                channelId: item.snippet?.videoOwnerChannelId ?? item.snippet?.channelId ?? ""
            )
        }
        return VideoPage(items: items, nextPageToken: response.nextPageToken)
    }

    // MARK: - Private helpers

    private func fetchChannel(query extra: [(String, String)]) async throws -> Channel {
        var query: [(String, String)] = [("part", "snippet,contentDetails")]
        query.append(contentsOf: extra)
        let response: ChannelListResponse = try await get("channels", query: query)
        guard let item = response.items.first else {
            throw YouTubeAPIError.channelNotFound
        }
        return Channel(
            id: item.id,
            title: item.snippet?.title ?? "(チャンネル名なし)",
            thumbnailURL: item.snippet?.thumbnails?.bestURL,
            uploadsPlaylistId: item.contentDetails?.relatedPlaylists?.uploads
        )
    }

    /// カスタムURL名から search.list で channelId を引く（quota 100）。
    private func searchChannelId(byName name: String) async throws -> String {
        let query: [(String, String)] = [
            ("part", "snippet"),
            ("type", "channel"),
            ("q", name),
            ("maxResults", "1"),
        ]
        let response: SearchListResponse = try await get("search", query: query)
        guard let id = response.items.first?.id?.channelId else {
            throw YouTubeAPIError.channelNotFound
        }
        return id
    }

    /// 共通のGETリクエスト（デコードまで行う）。
    private func get<T: Decodable>(_ path: String, query: [(String, String)]) async throws -> T {
        let data = try await getData(path, query: query)
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw YouTubeAPIError.decodingError
        }
    }

    /// 共通のGETリクエスト（生データ）。エラーを YouTubeAPIError にマップする。
    private func getData(_ path: String, query: [(String, String)]) async throws -> Data {
        let key = try apiKey()
        var comps = URLComponents(string: "\(baseURL)/\(path)")!
        comps.queryItems = query.map { URLQueryItem(name: $0.0, value: $0.1) }
            + [URLQueryItem(name: "key", value: key)]
        guard let url = comps.url else { throw YouTubeAPIError.invalidURL }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(from: url)
        } catch {
            throw YouTubeAPIError.networkError
        }

        guard let http = response as? HTTPURLResponse else {
            throw YouTubeAPIError.unknown
        }
        switch http.statusCode {
        case 200...299:
            break
        case 403:
            // quota 超過かどうかを本文から判定。
            if let body = String(data: data, encoding: .utf8),
               body.contains("quotaExceeded") || body.contains("dailyLimitExceeded") {
                throw YouTubeAPIError.quotaExceeded
            }
            throw YouTubeAPIError.unknown
        case 404:
            throw YouTubeAPIError.channelNotFound
        default:
            throw YouTubeAPIError.networkError
        }

        return data
    }
}

// MARK: - ISO8601 パース（fractional seconds 有無の両対応）

enum ISO8601 {
    private static let withFraction: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    private static let plain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func date(from string: String) -> Date? {
        withFraction.date(from: string) ?? plain.date(from: string)
    }
}

// MARK: - API レスポンス（Decodable）

private struct ChannelListResponse: Decodable {
    let items: [Item]
    struct Item: Decodable {
        let id: String
        let snippet: Snippet?
        let contentDetails: ContentDetails?
    }
    struct Snippet: Decodable {
        let title: String?
        let thumbnails: Thumbnails?
    }
    struct ContentDetails: Decodable {
        let relatedPlaylists: RelatedPlaylists?
    }
    struct RelatedPlaylists: Decodable {
        let uploads: String?
    }
}

/// videos.list（共有された動画URL → 投稿チャンネルの特定）用。
private struct VideoListResponse: Decodable {
    let items: [Item]
    struct Item: Decodable {
        let id: String?
        let snippet: Snippet?
    }
    struct Snippet: Decodable {
        let channelId: String?
        let channelTitle: String?
        let title: String?
    }
}

private struct SearchListResponse: Decodable {
    let items: [Item]
    struct Item: Decodable {
        let id: ID?
    }
    struct ID: Decodable {
        let channelId: String?
    }
}

private struct PlaylistItemListResponse: Decodable {
    let items: [Item]
    let nextPageToken: String?
    struct Item: Decodable {
        let snippet: Snippet?
        let contentDetails: ContentDetails?
    }
    struct Snippet: Decodable {
        let title: String?
        let description: String?
        let publishedAt: String?
        let channelId: String?
        let videoOwnerChannelId: String?
        let thumbnails: Thumbnails?
        let resourceId: ResourceId?
    }
    struct ResourceId: Decodable {
        let videoId: String?
    }
    struct ContentDetails: Decodable {
        let videoId: String?
        let videoPublishedAt: String?
    }
}

/// 各種 thumbnails オブジェクト（共通）。
private struct Thumbnails: Decodable {
    let `default`: Thumb?
    let medium: Thumb?
    let high: Thumb?
    let standard: Thumb?
    let maxres: Thumb?

    struct Thumb: Decodable {
        let url: String?
    }

    /// 利用可能な中で品質の高いサムネイルURL。
    var bestURL: URL? {
        let candidate = maxres?.url ?? standard?.url ?? high?.url ?? medium?.url ?? `default`?.url
        return candidate.flatMap(URL.init(string:))
    }
}

File: android/app/src/main/java/com/deskflowlabs/channeltimelineviewer/network/YouTubeApiClient.kt
package com.deskflowlabs.channeltimelineviewer.network

import android.net.Uri
import com.deskflowlabs.channeltimelineviewer.BuildConfig
import com.deskflowlabs.channeltimelineviewer.model.Channel
import com.deskflowlabs.channeltimelineviewer.model.VideoItem
import com.deskflowlabs.channeltimelineviewer.model.sortedByPublishedDate
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.time.Instant
import java.time.format.DateTimeParseException
import java.util.concurrent.TimeUnit

/** 動画一覧の1ページ分。 */
data class VideoPage(val items: List<VideoItem>, val nextPageToken: String?)

/**
 * YouTube Data API v3 クライアント。iOS 版 `Services/YouTubeAPIClient.swift` の移植。
 * スクレイピングは行わず、公式の Data API のみを使う。
 */
class YouTubeApiClient(
    private val apiKey: String = BuildConfig.YOUTUBE_API_KEY,
    private val httpClient: OkHttpClient = defaultClient,
    private val baseUrl: String = "https://www.googleapis.com/youtube/v3",
) {

    /** 暴走防止のための最大ページ数（50件/ページ × 100 = 5000本）。 */
    private val maxPages = 100

    /** 入力URL（または handle / channelId）からチャンネルを解決する。 */
    suspend fun resolveChannel(inputUrl: String): Channel =
        when (val identifier = ChannelResolver.parse(inputUrl)) {
            is ChannelIdentifier.ChannelId -> fetchChannel(listOf("id" to identifier.value))
            is ChannelIdentifier.Handle -> fetchChannel(listOf("forHandle" to "@${identifier.value}"))
            is ChannelIdentifier.Username -> fetchChannel(listOf("forUsername" to identifier.value))
            is ChannelIdentifier.CustomName ->
                fetchChannel(listOf("id" to searchChannelId(identifier.value)))
            // 共有された動画URL → videos.list で投稿チャンネルを特定してから解決する（quota 1）。
            is ChannelIdentifier.Video ->
                fetchChannel(listOf("id" to fetchChannelIdForVideo(identifier.videoId)))
        }

    /** videoId からその動画を投稿したチャンネルの channelId を取得する（videos.list / quota 1）。 */
    suspend fun fetchChannelIdForVideo(videoId: String): String {
        if (!ChannelResolver.isVideoId(videoId)) throw YouTubeApiException(YouTubeApiError.InvalidVideoUrl)
        val body = getJson("videos", listOf("part" to "snippet", "id" to videoId))
        return channelIdFromVideosList(body)
    }

    /** uploads プレイリストから全動画を取得し、古い順（publishedAt 昇順）で返す。 */
    suspend fun fetchVideos(playlistId: String): List<VideoItem> {
        val all = mutableListOf<VideoItem>()
        var token: String? = null
        var page = 0
        do {
            val result = fetchVideosPage(playlistId, token)
            all += result.items
            token = result.nextPageToken
            page += 1
        } while (token != null && page < maxPages)
        return all.sortedByPublishedDate(ascending = true)
    }

    /**
     * 既に持っている動画に当たるまで、新しい順にページを取得して**新着だけ**返す。
     *
     * @return 新着（新しい順）と、既知の動画に到達したかどうか。
     *   到達しなかった場合は差分が大きいので、呼び出し側で全件取得に切り替える。
     */
    suspend fun fetchNewVideos(
        playlistId: String,
        knownVideoIds: Set<String>,
        maxPages: Int = 5,
    ): Pair<List<VideoItem>, Boolean> {
        val newItems = mutableListOf<VideoItem>()
        var token: String? = null
        var page = 0
        var reachedKnown = false

        do {
            val result = fetchVideosPage(playlistId, token)
            for (item in result.items) {
                if (item.id in knownVideoIds) {
                    reachedKnown = true
                    break
                }
                newItems += item
            }
            if (reachedKnown) break
            token = result.nextPageToken
            page += 1
        } while (token != null && page < maxPages)

        // 最後まで見ても既知に当たらなかった＝そもそも全部が新しい（＝全件取得すべき）。
        return newItems to (reachedKnown || token == null)
    }

    /** uploads プレイリストの1ページ分を取得する。 */
    suspend fun fetchVideosPage(playlistId: String, pageToken: String?): VideoPage {
        val query = mutableListOf(
            "part" to "snippet,contentDetails",
            "playlistId" to playlistId,
            "maxResults" to "50",
        )
        if (pageToken != null) query += "pageToken" to pageToken

        val body = getJson("playlistItems", query)
        return videoPage(body)
    }

    private suspend fun fetchChannel(extraQuery: List<Pair<String, String>>): Channel {
        val body = getJson("channels", listOf("part" to "snippet,contentDetails") + extraQuery)
        val item = body["items"]?.jsonArray?.firstOrNull()?.jsonObject
            ?: throw YouTubeApiException(YouTubeApiError.ChannelNotFound)
        val snippet = item["snippet"]?.jsonObject
        return Channel(
            id = item.string("id") ?: throw YouTubeApiException(YouTubeApiError.ChannelNotFound),
            title = snippet?.string("title") ?: UNTITLED_CHANNEL,
            thumbnailUrl = bestThumbnail(snippet),
            uploadsPlaylistId = item["contentDetails"]?.jsonObject
                ?.get("relatedPlaylists")?.jsonObject?.string("uploads"),
        )
    }

    /** カスタムURL名から search.list で channelId を引く（quota 100）。 */
    private suspend fun searchChannelId(name: String): String {
        val body = getJson(
            "search",
            listOf("part" to "snippet", "type" to "channel", "q" to name, "maxResults" to "1"),
        )
        return body["items"]?.jsonArray?.firstOrNull()?.jsonObject
            ?.get("id")?.jsonObject?.string("channelId")
            ?: throw YouTubeApiException(YouTubeApiError.ChannelNotFound)
    }

    /** 共通のGETリクエスト。エラーを YouTubeApiError にマップする。 */
    private suspend fun getJson(path: String, query: List<Pair<String, String>>): JsonObject {
        if (apiKey.isBlank()) throw YouTubeApiException(YouTubeApiError.ApiKeyMissing)

        val builder = Uri.parse("$baseUrl/$path").buildUpon()
        query.forEach { (name, value) -> builder.appendQueryParameter(name, value) }
        builder.appendQueryParameter("key", apiKey)
        val url = builder.build().toString()

        val (code, body) = withContext(Dispatchers.IO) {
            try {
                httpClient.newCall(Request.Builder().url(url).get().build()).execute().use { response ->
                    response.code to (response.body?.string().orEmpty())
                }
            } catch (e: IOException) {
                throw YouTubeApiException(YouTubeApiError.NetworkError)
            }
        }

        when {
            code in 200..299 -> Unit
            code == 403 ->
                // quota 超過かどうかを本文から判定。
                if (body.contains("quotaExceeded") || body.contains("dailyLimitExceeded")) {
                    throw YouTubeApiException(YouTubeApiError.QuotaExceeded)
                } else {
                    throw YouTubeApiException(YouTubeApiError.Unknown)
                }
            code == 404 -> throw YouTubeApiException(YouTubeApiError.ChannelNotFound)
            else -> throw YouTubeApiException(YouTubeApiError.NetworkError)
        }

        return runCatching { json.parseToJsonElement(body).jsonObject }
            .getOrElse { throw YouTubeApiException(YouTubeApiError.DecodingError) }
    }

    companion object {
        /** タイトルが取れなかったときの表示。文言は画面側で置き換えられるよう素の文字列にする。 */
        const val UNTITLED_VIDEO = "(No title)"
        const val UNTITLED_CHANNEL = "(No channel name)"

        private val json = Json { ignoreUnknownKeys = true; isLenient = true }

        private val defaultClient: OkHttpClient by lazy {
            OkHttpClient.Builder()
                .connectTimeout(20, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build()
        }

        /** videos.list のレスポンスから channelId を取り出す（ネットワーク非依存＝テスト可能）。 */
        fun channelIdFromVideosList(body: JsonObject): String {
            val channelId = body["items"]?.jsonArray?.firstOrNull()?.jsonObject
                ?.get("snippet")?.jsonObject?.string("channelId")
            if (channelId.isNullOrEmpty()) throw YouTubeApiException(YouTubeApiError.VideoNotFound)
            return channelId
        }

        /** playlistItems のレスポンスを VideoItem に変換する（ネットワーク非依存＝テスト可能）。 */
        fun videoPage(body: JsonObject): VideoPage {
            val items = body["items"]?.jsonArray.orEmpty().mapNotNull { element ->
                val item = element.jsonObject
                val snippet = item["snippet"]?.jsonObject
                val contentDetails = item["contentDetails"]?.jsonObject
                val videoId = contentDetails?.string("videoId")
                    ?: snippet?.get("resourceId")?.jsonObject?.string("videoId")
                    ?: return@mapNotNull null
                val publishedText = contentDetails?.string("videoPublishedAt")
                    ?: snippet?.string("publishedAt")
                VideoItem(
                    id = videoId,
                    title = snippet?.string("title") ?: UNTITLED_VIDEO,
                    description = snippet?.string("description").orEmpty(),
                    publishedAtEpochSeconds = parseIso8601(publishedText),
                    thumbnailUrl = bestThumbnail(snippet),
                    channelId = snippet?.string("videoOwnerChannelId")
                        ?: snippet?.string("channelId").orEmpty(),
                )
            }
            return VideoPage(items, body.string("nextPageToken"))
        }

        /** 一番大きいサムネイルを選ぶ（maxres → standard → high → medium → default）。 */
        fun bestThumbnail(snippet: JsonObject?): String? {
            val thumbnails = snippet?.get("thumbnails")?.jsonObject ?: return null
            for (size in listOf("maxres", "standard", "high", "medium", "default")) {
                thumbnails[size]?.jsonObject?.string("url")?.let { return it }
            }
            return null
        }

        /** ISO8601（小数秒の有無どちらも）をエポック秒にする。読めなければ 0。 */
        fun parseIso8601(text: String?): Long {
            if (text.isNullOrBlank()) return 0
            return try {
                Instant.parse(text).epochSecond
            } catch (e: DateTimeParseException) {
                0
            }
        }

        private fun JsonObject.string(key: String): String? =
            runCatching { this[key]?.jsonPrimitive?.content }.getOrNull()
    }
}
